specs/agents/tracing-instrumentation-azure.md

```diff
@@ -0,0 +1,406 @@
+# Azure services spans
+
+We describe how to instrument some of Azure's services in this document.
+Some of the services can use existing specs. When there are differences or additions, they have been noted below.
+
+
+----
+
+**NOTE**: 
+Azure Storage can be run as part of [Azure Stack](https://azure.microsoft.com/en-au/overview/azure-stack/). Azure services cannot be inferred from Azure Stack HTTP service endpoints.
+
+----
+
+## Azure Storage
+
+Azure Storage is a collection of storage services, accessed through a
+Storage account. The services include
+
+- Blob storage
+- Queue storage
+- File share storage
+- Table storage
+
+A storage account is created in an Azure location (_region_), but the
+location is not discoverable through REST API calls.
+
+### Blob storage
+
+Object storage for binary and text data via a REST API. There are three types of blobs: block blobs, page blobs and append blobs.
+
+Blobs are organized into containers, where a container can contain an
+unlimited number of blobs, and a storage account can contain an unlimited
+number of containers. Although blob storage is a flat storage scheme
+(only one level of containers), blob names can include path segments (`/`),
+providing a _virtual_ hierarchy.
+
+A resource name is the full name of the container and blob. For example, Given container `foo` and blob `bar/baz`, the resource name
+is `foo/bar/baz`.
+
+| APM field | Required? | Format | Notes | Example |
+| --------- | --------- | ------ | ----- | ------- |
+| `span.name` | yes | `AzureBlob <OperationName> <ResourceName>` | Pascal case Operation name | `AzureBlob Upload foo/bar/baz` |
+| `span.type` | yes | `storage` | | |
+| `span.subtype` | yes | `azureblob` | | |
+| `span.action` | yes | `<OperationName>` | Pascal case | `Upload` |
+
+#### Span context fields
+
+| APM field | Required? | Format | Notes | Example |
+| --------- | --------- | ------ | ----- | ------- |
+| `context.destination.address` | yes | URL scheme and host | | `https://accountname.blob.core.windows.net/` |
+| `context.destination.service.name` | yes | `azureblob` | | | 
+| `context.destination.service.resource` | yes | `azureblob/<ResourceName>` | | `azureblob/foo/bar` |

```

Comment:
Having the blob as part of the resource name feels like it could have a too high cardinality for this field. Some background: The APM Server creating metrics based on spans and uses `context.destination.service.resource` as a dimension.
Another implication is that each distinct service resource results in a node on the service map.

What about setting `context.destination.service.resource` to `azureblob/${container}`?

Also, I'm unsure how much `context.destination.service.name` and `context.destination.service.resource` need to be in sync. Is the service.name a human-readable version for `service.resource`? @axw do you happen to know?

Reply:
They don't need to be the same: `name` is the human-readable string, and will be used for grouping in the UI; `resource` should represent how the source service accessed the destination.

(At least that's how I remember Ron explaining it :))

Reply:
> What about setting `context.destination.service.resource` to `azureblob/${container}`?

If the service resource is used in the service map, perhaps `context.destination.service.resource` set to `azure/${storage account name}/blob` might be a useful descriptor on which to group?

Reply:
Updated to use `azureblob/${storage account name}`